### PR TITLE
GitHub Integration (push files to github repo directly from ipynb file)

### PR DESCRIPTION
**Pull request by nuuron**

### CODE REVIEW

File: vevestaX/vevesta.py [hunk 1]
```diff
@@ -828,3 +828,34 @@ def commit(self, techniqueUsed, filename=None, message=None, version=None, proje
...
+    def git_commit(self, repoName, sourceBranch='main', newBranch=None, commitMessage=None):
+        if newBranch is None:
+            newBranch = sourceBranch
+
+        git_access_token = open('git_access_token.txt', "r").read()
+        g = Github(git_access_token)
+        repo = g.get_repo(repoName)
+        sb = repo.get_branch(sourceBranch)
+
+        ipynb_file_name = self.get_filename()
```
Comment (Priyanka-Vevesta) [inline vevestaX/vevesta.py:841]: there is a function that does that...please use its output

File: vevestaX/vevesta.py [hunk 2]
```diff
@@ -828,3 +828,34 @@ def commit(self, techniqueUsed, filename=None, message=None, version=None, proje
             print("Wrote experiment to tool, Vevesta")
         else:
             print("Failed to write experiment to tool, Vevesta")
+
+    def git_commit(self, repoName, sourceBranch='main', newBranch=None, commitMessage=None):
```
Comment (Priyanka-Vevesta) [inline vevestaX/vevesta.py:832]: git commit will be private function

Comment (Priyanka-Vevesta) [inline vevestaX/vevesta.py:832]: WHy is source branch required

File: vevestaX/vevesta.py (hunk 1 above)
Comment (Priyanka-Vevesta) [inline vevestaX/vevesta.py:841]: this will work for python code too.. i mean py files

File: vevestaX/vevesta.py [hunk 3]
```diff
@@ -828,3 +828,34 @@ def commit(self, techniqueUsed, filename=None, message=None, version=None, proje
...
+        repo = g.get_repo(repoName)
+        sb = repo.get_branch(sourceBranch)
+
+        ipynb_file_name = self.get_filename()
+        if os.path.exists(ipynb_file_name):
+            ipynb_file_content = open(ipynb_file_name, 'r').read()
+            try:
+                contents = repo.get_contents(ipynb_file_name, ref=sourceBranch)
+                if commitMessage is None:
+                    commitMessage = 'updated ' + ipynb_file_name
```
Comment (Priyanka-Vevesta) [inline vevestaX/vevesta.py:847]: the technique used in the commit function will be used to create the branch name and the message will be commit message

File: vevestaX/vevesta.py [hunk 4]
```diff
@@ -828,3 +828,34 @@ def commit(self, techniqueUsed, filename=None, message=None, version=None, proje
...
+                    contents.path,
+                    commitMessage,
+                    ipynb_file_content,
+                    sha=contents.sha,
+                    branch=newBranch
+                )
+                print(result)
+            except FileNotFoundError:
+                repo.create_git_ref(ref='refs/heads/' + newBranch, sha=sb.commit.sha)
+                if commitMessage is None:
```
Comment (Priyanka-Vevesta) [inline vevestaX/vevesta.py:858]: if message is none, the technique used will be the commit message.

**Review by Priyanka-Vevesta — CHANGES_REQUESTED** [review]

File: vevestaX/vevesta.py (hunk 2 above)
Comment (Priyanka-Vevesta) [reply vevestaX/vevesta.py:832]: git commit will be called from inside commit function

Comment (Priyanka-Vevesta) [reply vevestaX/vevesta.py:832]: It wont ask for  new branch name as argument, instead use the technique used appended with some 6 random numbers at the end.

File: requirements.txt [hunk 5]
```diff
@@ -7,3 +7,6 @@ pandas==1.4.2
 pyspark==3.2.1
 requests==2.27.1
 scipy==1.8.0
+
+setuptools~=60.2.0
+PyGithub~=1.55
```
Comment (Priyanka-Vevesta) [inline requirements.txt:12]: Please update setup.py too always

File: vevestaX/vevesta.py [hunk 6]
```diff
@@ -780,13 +803,34 @@ def __getExcelSheetData(self, fileName, sheetName):
                 modelingData = pandas.read_excel(fileName, sheet_name=sheetName, index_col=[])
                 return modelingData
 
-    def commit(self, techniqueUsed, filename=None, message=None, version=None, projectId=None, attachmentFlag=True):
+    def commit(self, techniqueUsed, filename=None, message=None, version=None, projectId=None, attachmentFlag=True,
+               repoName=None, sourceBranch=None, newBranch=None):
         self.dump(techniqueUsed, filename=filename, message=message, version=version, showMessage=False)
 
         # api-endpoint
```
Comment (Priyanka-Vevesta) [inline vevestaX/vevesta.py:810]: Please write this inside a private function

File: vevestaX/vevesta.py [hunk 7]
```diff
@@ -780,13 +803,34 @@ def __getExcelSheetData(self, fileName, sheetName):
...
+        params = {'projectId': projectId}
+
+        # get git access token
+        response = requests.post(url=backend_url + '/GetGitToken', headers=headers, params=params)
+        data = response.json()
+        git_token = data['gitToken']
+        project_title = data['projectTitle']
+        if repoName is None:
+            repoName = project_title
+        if sourceBranch is None:
```
Comment (Priyanka-Vevesta) [inline vevestaX/vevesta.py:824]: this isnt correct... source branch is master branch on which nothing will be pushed except when it has been specified also by the user as newbranch

**Review by Priyanka-Vevesta — CHANGES_REQUESTED** [review]

File: vevestaX/vevesta.py [hunk 8]
```diff
@@ -812,26 +856,23 @@ def commit(self, techniqueUsed, filename=None, message=None, version=None, proje
...
+            git_token = self.__find_git_token(backend_url=backend_url, access_token=token)
+            if repoName is None:
+                project = self.__fetch_project(backend_url=backend_url, access_token=token, projectId=projectId)
+                project_title = project['title']
+                repoName = project_title
+
+            if sourceBranch is None:
+                sourceBranch = techniqueUsed
+            if newBranch is None:
+                newBranch = sourceBranch
```
Comment (Priyanka-Vevesta) [inline vevestaX/vevesta.py:870]: new branch will be techniqueUsed...when source branch is main, then it that case a new branch say XGboost will be created

**Review by Priyanka-Vevesta — CHANGES_REQUESTED** [review]

**Review by jahnavisana — APPROVED** [review]

**Review by pavanVevesta — APPROVED** [review]

**Pull request merged**